### PR TITLE
Merge usac to 5.x

### PR DESCRIPTION
**Pull request by ivashmak**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

Base branch is PR #23979. Merging PR #23078 and PR #23806 to 5.x

### CODE REVIEW

File: modules/3d/src/usac.hpp [hunk 1]
```diff
@@ -130,19 +131,22 @@ class MinimalSolver : public Algorithm {
...
 //-------------------------- HOMOGRAPHY MATRIX -----------------------
-class HomographyMinimalSolver4ptsGEM : public MinimalSolver {
+class HomographyMinimalSolver4pts : public MinimalSolver {
 public:
-    static Ptr<HomographyMinimalSolver4ptsGEM> create(const Mat &points_);
+    static Ptr<HomographyMinimalSolver4pts> create(const Mat &points, bool use_ge);
+};
+class PnPSVDSolver : public MinimalSolver {
+public:
+    static Ptr<PnPSVDSolver> create (const Mat &points);
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:143]: there is PnPSVDSolver in 4.x, it's just on another line:
https://github.com/opencv/opencv/blob/4.x/modules/calib3d/src/usac.hpp#L97

File: modules/3d/src/usac/dls_solver.cpp [hunk 2]
```diff
@@ -46,29 +46,27 @@
...
+public:
+    explicit DLSPnPImpl (const Mat &, const Mat &, const Mat &) {}
 #endif
-        {}
+
     // return minimal sample size required for non-minimal estimation.
     int getMinimumRequiredSampleSize() const override { return 3; }
     // return maximum number of possible solutions.
     int getMaxNumberOfSolutions () const override { return 27; }
-    Ptr<NonMinimalSolver> clone () const override {
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac/dls_solver.cpp:69]: are you sure? In 4.x. there's no clone() method either:
https://github.com/opencv/opencv/blob/4.x/modules/calib3d/src/usac/dls_solver.cpp#L69

File: modules/3d/src/usac.hpp [hunk 3]
```diff
@@ -924,26 +1078,56 @@ Stage 1b, 2b, 3a, 3b, 5 may jump back to Stage 1a.
...
+    Ptr<WeightFunction> _weight_fnc;
+
+    int points_size, _state, filtered_points_size;
+    double threshold, max_thr;
+    bool parallel;
+
+    Matx33d T1, T2;
+    Mat points, K1, K2, calib_points, image_points, norm_points, filtered_points;
+    Ptr<NeighborhoodGraph> graph;
+    std::vector<Ptr<NeighborhoodGraph>> layers;
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:1104]: Yes, but the problem is that in 5.x those parameters are in usac.hpp, and they are used in other files, so I had to add more parameters in the end:
https://github.com/opencv/opencv/blob/5.x/modules/3d/src/usac.hpp#L926

File: modules/3d/src/usac.hpp [hunk 4]
```diff
@@ -549,6 +608,7 @@ class NeighborhoodGraph : public Algorithm {
     virtual ~NeighborhoodGraph() override = default;
     // Return neighbors of the point with index @point_idx_ in the graph.
     virtual const std::vector<int> &getNeighbors(int point_idx_) const = 0;
+    virtual const std::vector<std::vector<int>> &getGraph () const = 0;
 };
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:612]: It is not used, I forgot to remove it in 4.x in usac.hpp

File: modules/3d/src/usac.hpp [hunk 5]
```diff
@@ -6,34 +6,31 @@
...
-    virtual int getMaxNumHypothesisToTestBeforeRejection() const = 0;
     virtual int getRandomGeneratorState () const = 0;
     virtual bool isParallel() const = 0;
 
     virtual NeighborSearchMethod getNeighborsSearchMethod () const = 0;
     virtual SamplingMethod getSamplingMethod () const = 0;
     virtual ScoreMethod getScoreMethod () const = 0;
     virtual LocalOptimMethod getLOMethod () const = 0;
-
+    virtual EstimationMethod getEstimationMethod () const = 0;
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:26]: Yes, I had to added it to be consistent with 5.x

File: modules/3d/src/usac.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (asmorkalov) [reply modules/3d/src/usac.hpp:143]: Please use the same file structure and not swap classes to reduce merge conflicts.

File: modules/3d/src/usac.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (asmorkalov) [reply modules/3d/src/usac.hpp:612]: Please create PR to 4.x then and mention it in the description.

File: modules/3d/src/usac.hpp [hunk 6]
```diff
@@ -6,34 +6,31 @@
 #define OPENCV_USAC_USAC_HPP
 
 namespace cv { namespace usac {
-enum EstimationMethod { Homography, Fundamental, Fundamental8, Essential, Affine, P3P, P6P};
-enum VerificationMethod { NullVerifier, SprtVerifier };
-enum PolishingMethod { NonePolisher, LSQPolisher };
-enum ErrorMetric {DIST_TO_LINE, SAMPSON_ERR, SGD_ERR, SYMM_REPR_ERR, FORW_REPR_ERR, RERPOJ};
+enum EstimationMethod { HOMOGRAPHY=0, FUNDAMENTAL=1, FUNDAMENTAL8=2, ESSENTIAL=3, AFFINE=4, P3P=5, P6P=6, PLANE=7, SPHERE=8};
+enum VerificationMethod { NULL_VERIFIER=0, SPRT_VERIFIER=1, ASPRT=2 };
+enum ErrorMetric {DIST_TO_LINE, SAMPSON_ERR, SGD_ERR, SYMM_REPR_ERR, FORW_REPR_ERR, RERPOJ, POINT_TO_PLANE, POINT_TO_SPHERE};
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp]: fixed

File: modules/3d/src/usac.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:99]: fixed this by moving all new stuff created in 5.x to the end of the file.

File: modules/3d/src/usac.hpp [hunk 7]
```diff
@@ -173,6 +177,13 @@ class AffineMinimalSolver : public MinimalSolver {
     static Ptr<AffineMinimalSolver> create(const Mat &points_);
 };
 
+//---------------------- Trifocal tensor -------------------
+class TrifocalTensorMinimalSolver : public MinimalSolver {
+public:
+    static Ptr<TrifocalTensorMinimalSolver> create(const Mat &points_);
+    virtual void getFundamentalMatricesFromTensor (const cv::Mat &tensor, cv::Mat &F21, cv::Mat &F31) = 0;
+};
+
```
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp]: fixed

File: modules/3d/src/usac.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:524]: No, I created pull-request that removes unused features, see 24047

[the remark this answers is not in the document]
Comment (ivashmak) [reply modules/3d/src/usac.hpp:524]: See #24047

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**